src/parser.rs

```diff
@@ -2200,78 +2217,61 @@ impl<'input, 'abbrev, 'unit, Endian> EntriesCursor<'input, 'abbrev, 'unit, Endia
     ///     if delta_depth <= 0 {
     ///         break;
     ///     }
     ///
-    ///     let current = cursor.current()
-    ///         .expect("Should be at an entry")
-    ///         .expect("And we should parse the entry ok");
-    ///     first_entry_with_no_children = Some(current);
+    ///     first_entry_with_no_children = Some(current.clone());
     /// }
     ///
     /// println!("The first entry with no children is {:?}",
     ///          first_entry_with_no_children.unwrap());
     /// ```
-    pub fn next_dfs(&mut self) -> Option<isize> {
-        self.current_ref();
-        {
-            if self.cached_current.is_some() {
-                self.input = if let Some(after_attrs) = self.cached_current
-                    .as_ref()
-                    .unwrap()
-                    .after_attrs
-                    .get() {
-                    after_attrs
-                } else {
-                    for _ in self.cached_current.as_ref().unwrap().attrs() {
-                    }
-                    self.cached_current
-                        .as_ref()
-                        .unwrap()
-                        .after_attrs
-                        .get()
-                        .expect("should have after_attrs after iterating attrs")
-                };
-
-                let mut delta_depth =
-                    if self.cached_current.as_ref().unwrap().abbrev.has_children() {
-                        1
-                    } else {
-                        0
-                    };
+    pub fn next_dfs<'me>(&'me mut self)
+        -> ParseResult<Option<(isize,
+                               &'me DebuggingInformationEntry<'input, 'abbrev, 'unit, Endian>)>> {
+        let mut delta_depth = self.delta_depth;
+        loop {
+            // Keep eating null entries that mark the end of an entry's
+            // children.
+            let mut input = self.after_entry();
+            while input.len() > 0 && input[0] == 0 {
+                delta_depth -= 1;
+                input = &input[1..];
+            }
+            self.input = input;
+            self.cached_current = None;
 
-                // Keep eating null entries that mark the end of an entry's
-                // children.
-                while self.input.len() > 0 && self.input[0] == 0 {
-                    delta_depth -= 1;
-                    self.input = &self.input[1..];
+            // The next entry should be the one we want.
+            if try!(self.next_entry()).is_some() {
+                if let Some(ref entry) = self.cached_current {
+                    return Ok(Some((delta_depth, entry)));
                 }
 
-                self.cached_current = None;
-
-                if self.input.len() > 0 {
-                    Some(delta_depth)
-                } else {
-                    None
-                }
+                // Must have been a strange leb128 encoding of zero.
```

Comment:
I don't understand when this can happen -- if there are multiple 0 digits with the LEB128 continuation bit set? Eg the LEB128 equivalent of "00000000000"? If this is indeed the case that this code is handling, then can you add a more explicit comment about that? If not, then what is it?